src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs

```diff
@@ -194,25 +217,27 @@ public static class TextLoaderSaverCatalog
         /// A column may also have dense values followed by sparse values represented in this fashion. For example,
         /// a row containing "1 2 5 2:6 4:3" represents two dense columns with values 1 and 2, followed by 5 sparsely represented
         /// columns with values 0, 0, 6, 0, and 3. The indices of the sparse columns start from 0, even though 0 represents the third column.</param>
+        /// <param name="autoMapLoadColumns">Whether the members name of the class passed as type argument should be mapped to <see cref="LoadColumnNameAttribute"/>.
+        /// When <see langword="true"/> requires the <paramref name="hasHeader"/> parameter to be <see langword="true"/>.</param>
         /// <returns>The data view.</returns>
         public static IDataView LoadFromTextFile<TInput>(this DataOperationsCatalog catalog,
             string path,
             char separatorChar = TextLoader.Defaults.Separator,
             bool hasHeader = TextLoader.Defaults.HasHeader,
             bool allowQuoting = TextLoader.Defaults.AllowQuoting,
             bool trimWhitespace = TextLoader.Defaults.TrimWhitespace,
-            bool allowSparse = TextLoader.Defaults.AllowSparse)
+            bool allowSparse = TextLoader.Defaults.AllowSparse,
+            bool autoMapLoadColumns = TextLoader.Defaults.AutoMapLoadColumns)
         {
             Contracts.CheckNonEmpty(path, nameof(path));
             if (!File.Exists(path))
             {
                 throw Contracts.ExceptParam(nameof(path), "File does not exist at path: {0}", path);
             }
 
-            // REVIEW: it is almost always a mistake to have a 'trainable' text loader here.
-            // Therefore, we are going to disallow data sample.
+            var source = new MultiFileSource(path);
             return TextLoader.CreateTextLoader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar,
-                allowQuoting, allowSparse, trimWhitespace).Load(new MultiFileSource(path));
+                allowQuoting, allowSparse, trimWhitespace, autoMapLoadColumns, dataSample: source).Load(source);
         }
```

Comment:
The onnx related test failures you are seeing are a result of this line. 
When you supply a data sample and the file has a header, the header fields are added as annotations to columns which get saved in the onnx model when the pipeline is exported. Since the earlier models were created without the data sample, they don't have annotations and you are seeing failures when comparing against the baseline. 

It is debatable whether we should always add the column annotations when a header is present. There is a comment in the older version that explicitly disallows a data sample. This was prior to my time, so I am not sure what the background is on this. 

@eerhardt / @yaeldekel : Do either of you know why it is considered a mistake to have a 'trainable' text loader here and using a data sample?

@kevencarneiro : It is possible to take your PR and retain the old behavior if you change the lines above to the following:
```
            var source = new MultiFileSource(path);
            var dataSample = autoMapLoadColumns ? source : null;
            return TextLoader.CreateTextLoader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar,
                allowQuoting, allowSparse, trimWhitespace, autoMapLoadColumns, dataSample: dataSample).Load(source);
```

Reply:
Regarding the reason not to have a trainable text loader in the `LoadFromTextFile` API: When you create a trainable text loader, some parameters of the text loader are learned from the data sample. Specifically, slot names, if they exist, and in some cases the cardinality of some columns in the data (the range can be specified with only a minimum and no maximum, indicating the text loader needs to figure out the end of the range on its own). Lets imagine that a user loads a data view using `LoadFromTextFile`, and fits a pipeline using the data view to create a model. They then load new data using `LoadFromTextFile`, and try to apply the model to it. If the data view returned from `LoadFromTextFile` uses a trainable text loader, it could end up having different schemas depending on the actual data in the text file, which might cause the model to fail when applied to the second dataset. 



---
In reply to: [409767689](https://github.com/dotnet/machinelearning/pull/5021#discussion_r409767689) [](ancestors = 409767689)

Reply:
Given this, I think that the new `autoMapLoadColumns` option should not be present in `LoadFromTextFile`, only in the `CreateTextLoader` APIs.

---
In reply to: [411977600](https://github.com/dotnet/machinelearning/pull/5021#discussion_r411977600) [](ancestors = 411977600,409767689)